src/svc_auth_gss.c

```diff
@@ -582,24 +598,53 @@ _svcauth_gss(struct svc_req *req, bool *no_dispatch)
 		call_stat = svcauth_gss_validate(req, gd);
 		switch (call_stat) {
 		default:
+			mutex_unlock(&gd->lock);
+			gd_locked = false;
+			if (gd_hashed)
+				(void)authgss_ctx_hash_del(gd);
+			unref_svc_rpc_gss_data(gd);
 			svcauth_gss_return(RPCSEC_GSS_CREDPROBLEM);
 		case 0:
 			break;
 		}
 
-		if (!svcauth_gss_nextverf(req, gd, htonl(gc->gc_seq)))
+		if (!svcauth_gss_nextverf(req, gd, htonl(gc->gc_seq))) {
+			mutex_unlock(&gd->lock);
+			gd_locked = false;
+			if (gd_hashed)
+				(void)authgss_ctx_hash_del(gd);
+			unref_svc_rpc_gss_data(gd);
 			svcauth_gss_return(AUTH_FAILED);
+		}
 		break;
 
 	case RPCSEC_GSS_DESTROY:
-		if (req->rq_msg.cb_proc != NULLPROC)
+		if (req->rq_msg.cb_proc != NULLPROC) {
+			mutex_unlock(&gd->lock);
+			gd_locked = false;
+			if (gd_hashed)
+				(void)authgss_ctx_hash_del(gd);
+			unref_svc_rpc_gss_data(gd);
 			svcauth_gss_return(AUTH_FAILED);	/* XXX ? */
+		}
 
-		if (svcauth_gss_validate(req, gd))
+		if (svcauth_gss_validate(req, gd)) {
+			mutex_unlock(&gd->lock);
+			gd_locked = false;
+			if (gd_hashed)
+				(void)authgss_ctx_hash_del(gd);
+			unref_svc_rpc_gss_data(gd);
 			svcauth_gss_return(RPCSEC_GSS_CREDPROBLEM);
```

Comment:
can we fix svcauth_gss_return() to do the clean up in a failure case?  Ideally, we should remove the macros and replace it with a goto block.

Reply:
Yah, good suggestion. Will work on it.